sklearn/preprocessing/label.py

```diff
@@ -125,7 +125,7 @@ def transform(self, y):
         y : array-like of shape [n_samples]
         """
         check_is_fitted(self, 'classes_')
-        y = column_or_1d(y, warn=True)
+        y = np.array(column_or_1d(y, warn=True), dtype=self.classes_.dtype)
```

Comment:
Any specific reason for doing so? Maybe just treat empty array as a special case like below?

Reply:
If we do use `len(y) == 0`, this will allow some odd objects to pass. Like try this with `y = ''` in both cases,

with using `len(y) == 0` check and return:
```python
In [2]: le = LabelEncoder()

In [3]: le.fit(["a","b"])
Out[3]: LabelEncoder()

In [4]: le.transform('')
Out[4]: array([], dtype=float64)
```
which isn't expected I guess? Since the `fit` method explicitly states 
```python
"""
      Parameters
      ----------
        y : array-like of shape [n_samples]
           Target values.
"""
```

With the current use (i.e. dtype=self.classes_.dtype)
```python
In [6]: le = LabelEncoder()

In [7]: le.fit(["a","b"])
Out[7]: LabelEncoder()

In [8]: le.transform('')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-d47958d90147> in <module>()
----> 1 le.transform('')

/home/gxyd/dev/scikit-learn/sklearn/preprocessing/label.py in transform(self, y)
    126         """
    127         check_is_fitted(self, 'classes_')
--> 128         y = np.array(column_or_1d(y, warn=True), dtype=self.classes_.dtype)
    129 
    130         classes = np.unique(y)

/home/gxyd/dev/scikit-learn/sklearn/utils/validation.pyc in column_or_1d(y, warn)
    645         return np.ravel(y)
    646 
--> 647     raise ValueError("bad input shape {0}".format(shape))
    648 
    649 

ValueError: bad input shape ()
```
which I think is a reasonable error.

Reply:
Will it be better to use _num_samples(not len)? current code seems a bit magical, I can't find similar usage in the codebase.

Reply:
I was wrong that `len(y) == 0` (I had put this check before `column_or_1d(y, warn=True)`, it should be done after it)  check will be a problem.

I'm sorry about that.

But in any case use of `_num_samples` (rather than `len(y)`) seems more natural to me.

Reply:
Thanks @gxyd I've changed it :slightly_smiling_face:

Reply:
Please also use _num_samples below, I don't think len is a good choice.

Reply:
Please also use _num_samples here. Current solution seems a bit magical.

Reply:
Will do that

Reply:
Done!